modules/ml/src/main/java/org/apache/ignite/ml/inference/builder/IgniteDistributedInfModelBuilder.java

```diff
@@ -0,0 +1,356 @@
+    private static final String INFERENCE_SERVICE_NAME_PATTERN = "inference_service_%s";
+
+    /** Template of the inference request queue name. */
+    private static final String INFERENCE_REQ_QUEUE_NAME_PATTERN = "inference_queue_req_%s";
+
+    /** Template of the inference response queue name. */
+    private static final String INFERENCE_RES_QUEUE_NAME_PATTERN = "inference_queue_res_%s";
+
+    /** Default capacity for all queues used in this class (request queue, response queue, received queue). */
+    private static final int QUEUE_CAPACITY = 100;
+
+    /** Default configuration for Apache Ignite queues used in this class (request queue, response queue). */
+    private static final CollectionConfiguration queueCfg = new CollectionConfiguration();
+
+    /** Ignite instance. */
+    private final Ignite ignite;
+
+    /** Number of service instances maintaining to make distributed inference. */
+    private final int instances;
+
+    /**
+     * Constructs a new instance of Ignite distributed inference model builder.
+     *
+     * @param ignite Ignite instance.
+     * @param instances Number of service instances maintaining to make distributed inference.
+     */
+    public IgniteDistributedInfModelBuilder(Ignite ignite, int instances) {
+        this.ignite = ignite;
+        this.instances = instances;
+    }
+
+    /**
+     * Starts the specified in constructor number of service instances and request/response queues. Each service
+     * instance reads request queue, processes inbound requests and writes responses to response queue. The returned
+     * facade is represented by the {@link InfModel} operates with request/response queues, but hides these details
+     * behind {@link InfModel#predict(Object)} method of {@link InfModel}.
+     *
+     * Be aware that {@link InfModel#close()} method must be called to clear allocated resources, stop services and
+     * remove queues.
+     *
+     * @param reader Inference model reader.
+     * @param parser Inference model parser.
+     * @param <I> Type of model input.
+     * @param <O> Type of model output.
+     * @return Facade represented by {@link InfModel}.
+     */
+    @Override public <I extends Serializable, O extends Serializable> InfModel<I, Future<O>> build(
+        InfModelReader reader, InfModelParser<I, O> parser) {
+        return new DistributedInfModel<>(ignite, UUID.randomUUID().toString(), reader, parser, instances);
+    }
+
+    /**
+     * Apache Ignite service that makes inference reading requests from the request queue and writing responses to the
+     * response queue. This service is assumed to be deployed in {@link #build(InfModelReader, InfModelParser)} method
+     * and cancelled in {@link InfModel#close()} method of the inference model.
+     *
+     * @param <I> Type of model input.
+     * @param <O> Type of model output.
+     */
+    private static class IgniteDistributedInfModelService<I extends Serializable, O extends Serializable>
```

Comment:
Maybe you should place code lower than DistributedInfModel
just for reading comfort - firstly you use DistributedInfModel and write its definition, and secondly, using Service in InfModel write Service code after

Reply:
Done.